kernel/arch/dreamcast/hardware/sq.c

```diff
@@ -19,16 +19,43 @@
     Based on code by Marcus Comstedt, TapamN, and Moop
 */
 
+/** \brief   Store Queue 0 access register 
+    \ingroup store_queues
+*/
+#define QACR0 (*(volatile uint32_t *)(void *)0xff000038)
+
+/** \brief   Store Queue 1 access register 
+    \ingroup store_queues  
+*/
+#define QACR1 (*(volatile uint32_t *)(void *)0xff00003c)
+
+/** \brief   Set Store Queue QACR* registers
+    \ingroup store_queues
+*/
+#define SET_QACR_REGS(dest0, dest1) \
+    do { \
+        QACR0 = ((uintptr_t)(dest0)) >> 24; \
+        QACR1 = ((uintptr_t)(dest1)) >> 24; \
+    } while(0)
+
 static mutex_t sq_mutex = MUTEX_INITIALIZER;
 
-void sq_lock(void) {
+void sq_lock(void *dest0, void *dest1) {
     mutex_lock(&sq_mutex);
+    sq_wait();
+    SET_QACR_REGS(dest0, dest1);
 }
 
 void sq_unlock(void) {
     mutex_unlock(&sq_mutex);
 }
 
+void sq_wait(void) {
+    /* Wait for both store queues to complete */
+    uint32 *d = (uint32_t *)MEM_AREA_SQ_BASE;
```

Comment:
```suggestion
    uint32_t *d = (uint32_t *)MEM_AREA_SQ_BASE;
```

Reply:
I'll wait for more suggestions to accept all in a batch.